### PR TITLE
Qt build improvements and documentation

### PR DESCRIPTION
**Pull request by ntadej**
Various improvements to the Qt build:
 - always build core libraries as object libraries when building with Qt
 - build on Windows warningless
 - embed bitcode on iOS
 - fix windows symbol exports
 - add helper scripts (eventually to be used in the CI)

/cc @kiibimees @rinigus

### CODE REVIEW

**Comment by birkskyum** [comment]
This is really great work. I struggle to build and run from and m1 mac, so any help on build scripts are greatly appreciated. What I see when i run any of the four mac script os this, which doesn't have to be related to this pr even:
Screenshot 2022-04-27 at 13 46 43

**Comment by ntadej** [comment]
> This is really great work. I struggle to build and run from and m1 mac, so any help on build scripts are greatly appreciated. What I see when i run any of the four mac script os this, which doesn't have to be related to this pr even:
> Screenshot 2022-04-27 at 13 46 43

You need to add Qt to your path, e.g. `/Volumes/Extreme/Workspace/SDK/Qt/6.3.0/macos/bin` for Qt 6 or `/Volumes/Extreme/Workspace/SDK/Qt/5.15.2/clang_64/bin` for Qt 5 (replace the path with your Qt installation).

**Comment by birkskyum** [comment]
Great! I think the script runs through now for qt5 - how can I run the build?

**Comment by ntadej** [comment]
> Great! I think the script runs through now for qt5 - how can I run the build? I'm on a ferry, without connection, but I'll install and try qt6 today as well.

All the scripts use `ninja` as generator so you can just run `ninja` and it will build it.

**Comment by birkskyum** [comment]
In the readme we could add an example of where to put the terminal and run what. So what I do for instance is opening a terminal at the root of maplibre-gl-native, and then run './platform/qt/scripts/***.sh' followed by 'ninja'.

I'm still not completely sure how I can execute or test the build.

**Comment by ntadej** [comment]
I agree, will extend the README now.

**Comment by rinigus** [comment]
I am planning to test it a touch later. Would be interesting to see whether it still fits Sailfish OS and manual builds on Linux

**Comment by ntadej** [comment]
> I am planning to test it a touch later. Would be interesting to see whether it still fits Sailfish OS and manual builds on Linux

It may require a more recent version of CMake (although I did not explicitly increase the requirements). Would that be a problem?

**Comment by rinigus** [comment]
> It may require a more recent version of CMake (although I did not explicitly increase the requirements). Would that be a problem?

We will see. I hope that CMake version will not limit it much - would be pity to reduce compatibility because of cmake :)

**Comment by ntadej** [comment]
> > It may require a more recent version of CMake (although I did not explicitly increase the requirements). Would that be a problem?
> 
> We will see. I hope that CMake version will not limit it much - would be pity to reduce compatibility because of cmake :)

I'm happy to also setup a CI pipeline for Sailfish OS if it helps.

**Comment by birkskyum** [comment]
Screenshot 2022-04-28 at 15 46 49

I try to run the script, but I get this.. any ideas?

**Comment by ntadej** [comment]
> Screenshot 2022-04-28 at 15 46 49
> 
> I try to run the script, but I get this.. any ideas?

You chould actually set the installation prefix. Here the files will be "installed".

**Comment by birkskyum** [comment]
That is a very good point. I tried to just write something in the installation prefix, and now it compiles

**Comment by ntadej** [comment]
> That is a very good point. I tried to just write something in the installation prefix, and now it compiles

OK, will add additional wording to the README 😊

**Comment by birkskyum** [comment]
Got the mbgl-qt running now, and the mbgl-test-runner fails at this:

[ RUN      ] Annotations.SymbolAnnotation
[ INFO ]  "{unknown}[General]: GPU Identifier: Apple M1"

Also, I'm not completely sure when I used CMake if I'm currently using qt 5 or 6

**Comment by rinigus** [comment]
Adding Sailfish CI isn't trivial. We are currently using OBS for the builds and I am building for large number of SFOS versions/arch combinations over there. Some builds are at https://build.sailfishos.org/package/show/sailfishos:chum:testing/qmaplibregl

On my own repo, I just started tests with your PR. Will take a bit of time to complete...

**Comment by rinigus** [comment]
So, I am getting errors in the early build stages:

```
[   88s] + /usr/bin/cmake -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_INSTALL_PREFIX:PATH=/usr -DCMAKE_INSTALL_LIBDIR:PATH=/usr/lib -DINCLUDE_INSTALL_DIR:PATH=/usr/include -DLIB_INSTALL_DIR:PATH=/usr/lib -DSYSCONF_INSTALL_DIR:PATH=/etc -DSHARE_INSTALL_PREFIX:PATH=/usr/share -DCMAKE_SKIP_RPATH:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=ON -DMBGL_WITH_QT=ON -DMBGL_WITH_WERROR=OFF -DCMAKE_INSTALL_PREFIX:PATH=/usr -DMBGL_WITH_QT_HEADLESS=OFF -DMBGL_QT_LIBRARY_ONLY=ON -DMBGL_QT_STATIC=OFF -DMBGL_QT_WITH_INTERNAL_ICU=ON .
[   88s] -- The CXX compiler identification is GNU 8.3.0
[   88s] -- The C compiler identification is GNU 8.3.0
[   88s] -- Check for working CXX compiler: /usr/bin/c++
[   88s] -- Check for working CXX compiler: /usr/bin/c++ -- works
[   88s] -- Detecting CXX compiler ABI info
[   89s] -- Detecting CXX compiler ABI info - done
[   89s] -- Detecting CXX compile features
[   89s] -- Detecting CXX compile features - done
[   89s] -- Check for working C compiler: /usr/bin/cc
[   90s] -- Check for working C compiler: /usr/bin/cc -- works
[   90s] -- Detecting C compiler ABI info
[   90s] -- Detecting C compiler ABI info - done
[   90s] -- Detecting C compile features
[   90s] -- Detecting C compile features - done
[   90s] -- Configuring GL-Native with OpenGL renderer backend
[   90s] CMake Error at vendor/csscolorparser.cmake:16 (target_link_libraries):
[   90s]   Object library target "mbgl-vendor-csscolorparser" may not link to
[   90s]   anything.
[   90s] Call Stack (most recent call first):
[   90s]   CMakeLists.txt:987 (include)
[   90s] 
[   90s] 
[   90s] -- Configuring incomplete, errors occurred!
```

This is for CMake `cmake-3.11.4+git4`. Earlier, it used to work.

**Comment by ntadej** [comment]
According to this ticket, CMake 3.12 may be required: https://gitlab.kitware.com/cmake/cmake/-/issues/18010
The current minimum version set is 3.10.

I think I remember now that we had the same issue in the past. One option is to add another patch file to the build to allow the upstream configuration to be nicer, what do you think?

**Comment by ntadej** [comment]
> Got the mbgl-qt running now, which is awesome, and the mbgl-test-runner fails at this:
> 
> [ RUN ] Annotations.SymbolAnnotation [ INFO ] "{unknown}[General]: GPU Identifier: Apple M1" zsh: segmentation fault ./mbgl-test-runner
> 
> Also, I'm not sure how to select Qt5 or Qt6 now, because I didn't specify the build script like before.
> 
> I can see from CMakeCache.txt that Qt5 was used in this case by default

If you build on macOS then Qt version should be the one you added to the path. While tests build, they may not run yet.

**Comment by rinigus** [comment]
> I think I remember now that we had the same issue in the past. One option is to add another patch file to the build to allow the upstream configuration to be nicer, what do you think?

I am not very familiar with CMake and cannot judge on what's the best way. Can you tell me what I am expected to patch out? There is one more issue with the build, I'll try to look into it tomorrow and then give proper feedback

File: platform/qt/src/timer.cpp [hunk 1]
```diff
@@ -18,7 +18,7 @@ void Timer::Impl::start(uint64_t timeout, uint64_t repeat_, std::function<void (
     callback = std::move(cb);
 
     timer.setSingleShot(true);
-    timer.start(timeout);
+    timer.start(static_cast<std::chrono::milliseconds>(timeout));
```
Comment (rinigus) [inline platform/qt/src/timer.cpp:21]: This line fails on SFOS with error 

```
[ 1011s] /home/abuild/rpmbuild/BUILD/qmaplibregl-0.0.220220.3+ntadej.qt.improvements.20220428192425.1.gb8426c6/maplibre-gl-native/platform/qt/src/timer.cpp:21:64: error: no matching function for call to 'QTimer::start(std::chrono::milliseconds)'
[ 1011s]      timer.start(static_cast<std::chrono::milliseconds>(timeout));
[ 1011s]                                                                 ^
[ 1011s] In file included from /usr/include/qt5/QtCore/QTimer:1,
[ 1011s]                  from /home/abuild/rpmbuild/BUILD/qmaplibregl-0.0.220220.3+ntadej.qt.improvements.20220428192425.1.gb8426c6/maplibre-gl-native/platform/qt/src/timer_impl.hpp:6,
[ 1011s]                  from /home/abuild/rpmbuild/BUILD/qmaplibregl-0.0.220220.3+ntadej.qt.improvements.20220428192425.1.gb8426c6/maplibre-gl-native/platform/qt/src/timer.cpp:1:
[ 1011s] /usr/include/qt5/QtCore/qtimer.h:142:10: note: candidate: 'void QTimer::start(int)'
[ 1011s]      void start(int msec);
[ 1011s]           ^~~~~
[ 1011s] /usr/include/qt5/QtCore/qtimer.h:142:10: note:   no known conversion for argument 1 from 'std::chrono::milliseconds' {aka 'std::chrono::duration<long long int, std::ratio<1, 1000> >'} to 'int'
[ 1011s] /usr/include/qt5/QtCore/qtimer.h:144:10: note: candidate: 'void QTimer::start()'
[ 1011s]      void start();
[ 1011s]           ^~~~~
[ 1011s] /usr/include/qt5/QtCore/qtimer.h:144:10: note:   candidate expects 0 arguments, 1 provided
```

Same goes for a line 31 below. 

What is this conversion designed to achieve? Isn't timeout in ms already?

Comment (rinigus) [reply platform/qt/src/timer.cpp:21]: As a comment: we are on Qt 5.6 due to license changes in the latter Qt releases. So far, it hasn't been an issue for Mapbox/Maplibre Qt Native

**Comment by rinigus** [comment]
One more issue: looks like `/usr/include/mbgl` is not installed as a part of `make install`. Which would make it impossible to use qmapboxgl as a library - something we can do now and what was a purpose of earlier CMake scripts development.

Next, after patching your CMake changes with https://github.com/rinigus/maplibre-gl-native/commit/534243a80a78dda02c91d839bf3f742161420dab, I am still getting errors, but now something else than before:

```
[  270s] CMake Error at platform/qt/qt.cmake:147 (target_link_libraries):
[  270s]   Error evaluating generator expression:
[  270s] 
[  270s]     $<PLATFORM_ID:Windows,Emscripten>
[  270s] 
[  270s]   $<PLATFORM_ID> expression requires one or zero parameters.
[  270s] Call Stack (most recent call first):
[  270s]   CMakeLists.txt:1089 (include)
[  270s] 
[  270s] 
[  270s] CMake Error:
[  270s]   Error evaluating generator expression:
[  270s] 
[  270s]     $<PLATFORM_ID:Windows,Emscripten>
[  270s] 
[  270s]   $<PLATFORM_ID> expression requires one or zero parameters.
[  270s] 
[  270s] 
[  270s] CMake Error at platform/qt/qt.cmake:147 (target_link_libraries):
[  270s]   Error evaluating generator expression:
[  270s] 
[  270s]     $<PLATFORM_ID:Windows,Emscripten>
[  270s] 
[  270s]   $<PLATFORM_ID> expression requires one or zero parameters.
[  270s] Call Stack (most recent call first):
[  270s]   CMakeLists.txt:1089 (include)
[  270s] 
[  270s] 
[  270s] CMake Error:
[  270s]   Error evaluating generator expression:
[  270s] 
[  270s]     $<PLATFORM_ID:Windows,Emscripten>
[  270s] 
[  270s]   $<PLATFORM_ID> expression requires one or zero parameters.
[  270s] 
[  270s] 
[  270s] CMake Error at test/CMakeLists.txt:214 (target_link_libraries):
[  270s]   Error evaluating generator expression:
[  270s] 
[  270s]     $<PLATFORM_ID:Windows,Emscripten>
[  270s] 
[  270s]   $<PLATFORM_ID> expression requires one or zero parameters.
[  270s] 
[  270s] 
[  270s] CMake Error at benchmark/CMakeLists.txt:57 (target_link_libraries):
[  270s]   Error evaluating generator expression:
[  270s] 
[  270s]     $<PLATFORM_ID:Windows,Emscripten>
[  270s] 
[  270s]   $<PLATFORM_ID> expression requires one or zero parameters.
[  270s] 
[  270s] 
[  270s] CMake Error:
[  270s]   Error evaluating generator expression:
[  270s] 
[  270s]     $<PLATFORM_ID:Windows,Emscripten>
[  270s] 
[  270s]   $<PLATFORM_ID> expression requires one or zero parameters.
[  270s] 
[  270s] 
[  271s] CMake Error:
[  271s]   Error evaluating generator expression:
[  271s] 
[  271s]     $<PLATFORM_ID:Windows,Emscripten>
[  271s] 
[  271s]   $<PLATFORM_ID> expression requires one or zero parameters.
[  271s] 
[  271s] 
[  271s] -- Generating done
[  271s] CMake Warning:
[  271s]   Manually-specified variables were not used by the project:
[  271s] 
[  271s]     INCLUDE_INSTALL_DIR
[  271s]     LIB_INSTALL_DIR
[  271s]     LIB_SUFFIX
[  271s]     SHARE_INSTALL_PREFIX
[  271s]     SYSCONF_INSTALL_DIR
[  271s] 
[  271s] 
[  271s] -- Build files have been written to: /home/abuild/rpmbuild/BUILD/qmaplibregl-0.0.220220.3+ntadej.qt.improvements.20220430135841.2.gea101fb/maplibre-gl-native
[  271s] error: Bad exit status from /var/tmp/rpm-tmp.jChfv4 (%build)
```

So, to sum up:
- Qt compatibility was reduced by timer.cpp changes
- CMake is broken now for CMake 3.10 and I couldn't find solution for resolving it

File: platform/qt/src/timer.cpp (hunk 1 above)
Comment (ntadej) [reply platform/qt/src/timer.cpp:21]: OK, so we can set Qt 5.6 and CMake 3.10 as minimum supported for now. Will setup a CI pipeline.

Comment (ntadej) [reply platform/qt/src/timer.cpp:22]: It seems Qt changed the interface from `int` to `std::chronology::milliseconds` in version 5.8, fixed.

**Comment by ntadej** [comment]
Hi @rinigus, thanks for testing! Unfortunately Qt 5.6 is no longer available for install from Qt directly I may make a docker container with it. In the meantime I added a CI pipeline with CMake 3.10 to make sure this still works.

Hopefully all issues should be fixed now besides the include folder. Which headers from `mbgl` are actually needed as I do not see any explicit usage. I suspect `mbgl::TileServerOptions`?

P.S. What is your opinion on renaming the Qt bindings to `QMapLibre` for consistency?

**Comment by rinigus** [comment]
Re Qt 5.6: no wonder. If you could make a docker with it, that would be helpful

Re include files: without them it is harder to develop against qmaplibregl. TileServerOptions is obvious, but there are also constants, such as `mbgl::util::MIN_ZOOM`. So far it was that, I haven't used anything else in [mapbox-gl-qml](https://github.com/rinigus/mapbox-gl-qml)

Re QMapLibre: Yes, why not. it would require some packaging adjustments, but it makes sense. Although, it should be done with changing headers as well (qmapbox.hpp and friends). Hmm, maybe too much work for little gain?

**Comment by rinigus** [comment]
PS: But why not to install mbgl headers for now?

**Comment by ntadej** [comment]
Yes, I will restore the headers for now. I disabled them as they seemed to be included by accident.

**Comment by rinigus** [comment]
Please let me know when I should test it against SFOS versions. Thank you for working on it!

**Comment by ntadej** [comment]
> Please let me know when I should test it against SFOS versions. Thank you for working on it!

Once the CI is running I need to test the headers and then it will be ready, probably tomorrow

**Comment by birkskyum** [comment]
Anything holding this back?

**Comment by ntadej** [comment]
> Anything holding this back?

Yes, I need to finalize Linux CI and revert some changes for backwards compatibility.

**Comment by rinigus** [comment]
@ntadej: let me know when to test SFOS builds as well

**Comment by kiibimees** [comment]
Sorry for the delay. I very much support building the libs always as OBJECT when building for Qt. Makes the whole lib file management (install/deploy/link) simpler. Perhaps we could go even further and detect STATIC automatically like it's done in Qt:
```get_target_property(QtCoreLibraryType Qt${QT_VERSION_MAJOR}::Core TYPE)
if(${QtCoreLibraryType} MATCHES STATIC_LIBRARY)
   ...
endid()```

Because the static is useful only in two cases, yes?:
* When trying to produce single target lib file in older CMake.
* When building with static version of Qt (for example iOS).

But, just a proposal. No need to take it too seriously.

**Comment by ntadej** [comment]
Hi @rinigus, @birkskyum, this should be ready to be merged now.

@kiibimees, thank you for your proposal, I'll investigate it (but for a follow-up MR).

**Comment by rinigus** [comment]
Thanks! I will test the builds on SFOS. Will probably be able to report back tomorrow

**Comment by rinigus** [comment]
@ntadej: looks to be compiling nicely on SFOS versions that were working earlier. Thank you!

File: include/mbgl/actor/actor_ref.hpp [hunk 2]
```diff
@@ -38,7 +38,11 @@ class ActorRef {
     template <typename Fn, class... Args>
     auto ask(Fn fn, Args&&... args) const {
         // Result type is deduced from the function's return type
+#if __cplusplus >= 201703L  // TODO: remove this once fully migrated to C++17
```
Comment (wipfli) [inline include/mbgl/actor/actor_ref.hpp:41]:
```suggestion
#if __cplusplus >= 201703L
```

Comment (wipfli) [reply include/mbgl/actor/actor_ref.hpp:41]: I removed the todo and opened an issue instead to keep track of post c++ 17 clean-up tasks. https://github.com/maplibre/maplibre-gl-native/issues/335

**Review by wipfli — APPROVED** [review]
Thanks @ntadej and every one who helped moving this forward!
